sdk/typespec/typespec_derive/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+        // If there were any errors parsing attrs, emit their compile_error! statements and return.
+        let errs = context.errors();
+        return quote::quote! {
+            #(#errs)*
+        };
+    }
+
+    let format = attrs.format.unwrap_or(Format::Json);
+    let deserialize_body = match format {
+        Format::Json => quote::quote! {
+            body.json().await
+        },
+        Format::Xml => quote::quote! {
+            body.xml().await
+        },
+    };
+
+    // If the standard path is used, we need to add 'extern crate', because it's possible the calling code
+    // depends on typespec_client_core transitively, which means it's not in scope by default.
+    // That's not necessary when using a custom path because we assume the user has done that work.
+    let typespec_import = match attrs.typespec_path {
+        Some(path) => quote::quote! {
+            use #path as _typespec;
+        },
+        None => quote::quote! {
+            #[allow(unused_extern_crates, clippy::useless_attribute)]
+            extern crate typespec_client_core as _typespec;
+        },
+    };
+
+    quote::quote! {
+        #typespec_import
+
+        #[automatically_derived]
+        impl #impl_generics _typespec::http::Model for #name #ty_generics #where_clause {
+            async fn from_response_body(body: _typespec::http::ResponseBody) -> _typespec::Result<Self> {
+                #deserialize_body
+            }
+        }
+    }
+}
+
+enum Format {
+    Json,
+    Xml,
+}
+
+struct Context {
+    errors: Vec<TokenStream>,
+}
+
+impl Context {
+    pub fn new() -> Context {
+        Context { errors: Vec::new() }
+    }
+
+    pub fn emit_error(&mut self, span: Span, message: &str) {
+        let token_stream = quote::quote_spanned!(span=> compile_error!(#message));
+        self.errors.push(token_stream);
+    }
+
+    pub fn errors(&self) -> &[TokenStream] {
+        &self.errors
+    }
+}
+
+struct Attrs {
+    pub typespec_path: Option<Path>,
+    pub format: Option<Format>,
+}
+
+impl Attrs {
+    pub fn from_attrs(attrs: &[Attribute], ctx: &mut Context) -> Attrs {
+        let mut typespec_path = None;
+        let mut format = None;
+        for attr in attrs.iter().filter(|a| a.path().is_ident("typespec")) {
+            let Meta::List(meta_list) = &attr.meta else {
+                ctx.emit_error(attr.span(), "Invalid typespec attribute, expected attribute in form #[typespec(key = value)]");
+                continue;
+            };
+
+            meta_list.parse_nested_meta(|meta| {
+                let Some(ident) = meta.path.get_ident() else {
+                    ctx.emit_error(meta.path.span(), "Invalid typespec attribute, expected attribute in form #[typespec(key = value)]");
+                    return Ok(());
+                };
+                match ident.to_string().as_str() {
+                    "crate" => {
+                        let value = meta.value().unwrap();
```

Comment:
How likely is this to panic? Some unwraps in macros isn't rare, but seems a better error message is more informative if this fails less than rarely. My experience writing macros hasn't given me enough evidence to base any informed opinion here, so merely raising the question.

Reply:
Excellent catch, this fails when `#[typespec(whatever` is not followed by an `=` token. So, for example, `#[typespec(format("json"))]` (a potential misunderstanding of our attribute syntax) could cause it and we can provide a better error. I've updated this code and added an error test. The `value` is needed in both arms of the match, so I've gone ahead and promoted it above the `match`. That does change the behavior if you have both a malformed attribute, **and** reference an unknown attribute name (for example `#[typespec(not_something_we_support("bar"))]` would previously have errored with `unknown typespec attribute` and will now error with `invalid typespec attribute`). That seems reasonable to me though.